source/Nevermore/Advanced/ReadTransaction.cs

```diff
@@ -26,16 +26,29 @@ namespace Nevermore.Advanced
     public class ReadTransaction : IReadTransaction, ITransactionDiagnostic
     {
         static readonly ILog Log = LogProvider.For<ReadTransaction>();
-        static DbConnection DefaultConnectionFactory(string connectionString) => new SqlConnection(connectionString);
+        protected static DbConnection DefaultConnectionFactory(string connectionString) => new SqlConnection(connectionString);
 
-        readonly RelationalTransactionRegistry registry;
+        readonly IRelationalTransactionRegistry registry;
         readonly RetriableOperation operationsToRetry;
         readonly IRelationalStoreConfiguration configuration;
         readonly ITableAliasGenerator tableAliasGenerator = new TableAliasGenerator();
 
         readonly Func<string, DbConnection> connectionFactory;
+        readonly Action<string>? customCommandTrace;
         readonly string name;
 
+        protected abstract record ExistingConnectionAndTransaction(IConnectionAndTransaction Value)
```

Comment:
An [ADT](https://en.wikipedia.org/wiki/Algebraic_data_type) where all cases hold the same value feels a bit odd, maybe it's not the best tool for the job here?

The fact that the class can hold both an `ExistingConnectionAndTransaction` _and_ ` DbConnection` at the same time feels a bit weird too. It allows for this to hold two transactions at once which doesn't seem quite right?

Looking at the code, could we make it more idiomatic C# with something like the following:
It might clean things up a little bit too?

- Make a nevermore-internal class NevermoreConnectionAndTransaction : IConnectionAndTransaction to hold the places where Nevermore creates the transaction itself
- ReadTransaction always holds an IConnectionAndTransaction, rather than the nullable "existing" thing.
- This is either the nevermore-internal one or something else (who cares what type it is, doesn't matter)
- there's a `readonly bool ownsTransaction` flag that controls whether we dispose it here or not.

This is probably more in-line with what people would expect to see based on other code like Stream in C# which also has the "sometimes-owned" behaviour

Reply:
> An [ADT](https://en.wikipedia.org/wiki/Algebraic_data_type) where all cases hold the same value feels a bit odd, maybe it's not the best tool for the job here?

This was just personal preference, it offers no real advantage; it's functionally identical to something that carries an `IConnectionAndTransaction` along with an `enum Kind { ... }` for example but I just prefer to use the type itself to carry that info if I can. I'd be fine with changing it if there's strong disagreement.

> The fact that the class can hold both an ExistingConnectionAndTransaction and  DbConnection at the same time feels a bit weird too. It allows for this to hold two transactions at once which doesn't seem quite right?
>
> Looking at the code, could we make it more idiomatic C# with something like the following:
It might clean things up a little bit too?

I initially went that route but decided to take this approach purely for the sake of minimising changes. In this PR I was pretty torn between architecting things nicely vs a minimal change approach just to serve our needs. I'd be happy to change it and go that route after all if we decide it's worthwhile to invest in Nevermore like that, I think it's worth a discussion.

Reply:
> it's functionally identical to something that carries an IConnectionAndTransaction along with an enum Kind { ... }

Yep I am well acquainted with this pattern. 

The main thing that stuck out was the fact that the class holds both a `DbConnection` and a `ConnectionAndTransaction` member variable so can potentially hold two different transactions at once, relying on the correctness of code in various other places to deal with this correctly. This kind of code is IME, a common source of bugs.

While I agree with the sentiment of minimal-impact, and in the long term we want to move away from nevermore, I don't think that's going to come quite as soon as we might like. Minimising hacks in the nevermore codebase I think will still hold value for some time.

If it's not too much work, personally I'd like to see this cleaned up a little, but it's not a deal breaker. Defer to @emilol for a second opinion here.

Reply:
> In this PR I was pretty torn between architecting things nicely vs a minimal change approach just to serve our needs.

There's definitely a healthy tension there. Two goals that minimising changes can serve us in a public library that others depend on like this one can be: reducing the chance of a breaking change (not relevant here, we already have a breaking change) else, minimising what needs to change when upgrading. I don't think we get the latter either in this case? 

Minimising changes also comes with the risk of introducing bugs, but as we've already said in this thread, there's also a risk of bugs in this minimal approach. 

I'd be in favour of tidying this one up a bit. I haven't had a chance to look through the changes in detail, but I do wonder if there needs to be a nevermore-specific representation of a dbConnection and/or transaction, other than what we already have. Maybe this would be worth a mob after standup?

Reply:
No worries, I don't think there's any need to discuss further if it's worth the investment; I'll have another pass over this and go back to the approach I had originally that eliminates the redundant connection/transaction members.